analyzers/src/SonarAnalyzer.CSharp/Rules/AvoidUnderPosting.cs

```diff
@@ -0,0 +1,85 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class AvoidUnderPosting : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6964";
+    private const string MessageFormat = "Property used as input in a controller action should be nullable or annotated with the Required attribute to avoid under-posting.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStart =>
+        {
+            if (!compilationStart.Compilation.ReferencesControllers())
+            {
+                return;
+            }
+            var actionParameters = new ConcurrentStack<IParameterSymbol>();
+
+            compilationStart.RegisterSymbolStartAction(symbolStart =>
+            {
+                var method = (IMethodSymbol)symbolStart.Symbol;
+                if (method.IsControllerMethod())
+                {
+                    actionParameters.PushRange([.. method.Parameters]);
+                }
+            }, SymbolKind.Method);
```

Comment:
I would register for NamedType instead. The reason is, that we are only interested in MethodDecarations in Controller types. There is no need to register for all method declarations.
In the NamedType I would
* Check for "IsController", if yes
* Register for SyntaxNodes of type MethodDeclaration (All ASP.NET Actions must be "normal" methods. You could also register for ParameterDeclarations, but these would also fire for local functions and lambda and we are not interested in these)
* In the MethodDeclaration Action, I would query the ITypeSymbol for the parameter (as you did), and then look if it is defined in source (by checking DeclaringSyntaxReferences.Any())
* You also need to exclude some types that are understood by the MVC binder, like `IFormCollection` or `IIFormFile` or `IFormFileCollection`.
* The remaining logic you have looks good, but you may also need to look in base classes for properties.